sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java

```diff
@@ -53,19 +53,47 @@ public void skip() {
     throw new UnsupportedOperationException();
   }
 
+  private void updateCurrentByte() {
+    try {
+      currentByte = (byte) in.read();
+    } catch (IOException e) {
+      throw new ParquetDecodingException("Failed to read a byte", e);
+    }
+  }
+
   @Override
   public final void readBooleans(int total, WritableColumnVector c, int rowId) {
-    // TODO: properly vectorize this
-    for (int i = 0; i < total; i++) {
-      c.putBoolean(rowId + i, readBoolean());
+    int i = 0;
+    if (bitOffset > 0) {
+      i = Math.min(8 - bitOffset, total);
+      c.putBooleans(rowId, i, currentByte, bitOffset);
+      bitOffset = (bitOffset + i) & 7;
+    }
+    for (; i + 7 < total; i += 8) {
+      updateCurrentByte();
+      c.putBooleans(rowId + i, currentByte);
+    }
+    if (i < total) {
+      updateCurrentByte();
+      bitOffset = total - i;
+      c.putBooleans(rowId + i, bitOffset, currentByte, 0);
     }
   }
 
   @Override
   public final void skipBooleans(int total) {
-    // TODO: properly vectorize this
-    for (int i = 0; i < total; i++) {
-      readBoolean();
+    // using >>3 instead of /8 below since Java division rounds towards zero i.e. (-1)/8=0
```

Comment:
`(total - (8 - bitOffset))` can be negative.
E.g. `total=1`, `bitOffset=1`, for this combination, `(total - (8 - bitOffset))=(1-(8-1))=-6`. And `(-7)>>3=(-1)` is what we need. So the comment still holds. This example is covered in `ColumnarBatchSuite.scala`

Added `skipBooleans(0)` test.

Reply:
I should have clarified my comment - total should be larger than 8-bitOffset to skip at least 1 byte. Thanks for pointing it out.

What I meant was that comment does not reflect the actual purpose of bit shift and is confusing. All you are trying to figure out is whether numBytesToSkip is positive or not so both approaches would work just fine. Bit shift tends to be faster than the division operation - I am not sure if JVM is smart enough to realise that these operations are equivalent in this case. 

By the way, shouldn't 
```
if (bitOffset > 0) {
  updateCurrentByte();
}
```
be outside of the `if (numBytesToSkip >= 0) {`? In other words, what happens when the total is 9 and bitOffset is 0?

Reply:
Oh I see. The difference is important when `(total-(8-bitOffset))<0.` E.g. `(-1)>>3=(-1)` vs. `(-1)/8=0`. The latter incorrectly enters the `if(numBytesToSkip>=0){`. I updated the comment, hopefully it is clearer now.

We should not call
```
if (bitOffset > 0) {
  updateCurrentByte();
}
```
outside of the `if (numBytesToSkip >= 0) {` clause. That is because `numBytesToSkip<0` <=> `(total-(8-bitOffset))<0` means there will be still unread bits left in `currentByte`

Reply:
Hmm.. Why do you even need to go into the if statement if numBytesToSkip is 0?

Reply:
My concern is that the testing coverage is low and we could introduce subtle bugs that could be difficult to debug. I guess it is fine to keep as is.

Reply:
Thanks. For the record, I will try to explain here.

Let's say `total=8`, `bitOffset=1`, then there are `(8-bitOffset)=7` bits to skip in the `currentByte`. Now there is still 1 more bit to skip as `total=8`. So `updateCurrentByte()` needs to be called to update the `currentByte` and  `bitOffset` will be again `1`. In the future, the rest of the 7bits may be read from the updated `currentByte`. For that reason, we need to go into the if statement when `numBytesToSkip = (8 - (8 - 1)) >> 3 = 1>>3 = 0`.

The following is a few-lines longer but equivalent condition.
```
    if (numBytesToSkip > 0) {
      try {
        in.skipFully(numBytesToSkip);
      } catch (IOException e) {...}
    }
    if (numBytesToSkip >= 0 && bitOffset > 0) {
        updateCurrentByte();
    }
```

The scenario is tested at
https://github.com/apache/spark/pull/34611/files#diff-b84cbbb2eadfa9d267b9ab8be2e6be579f28c1813623785c9e667a864f7960e1R194